design/client-initiated-backchannel-authentication-flow.md

```diff
@@ -0,0 +1,494 @@
+
+### IDs for Identifying CIBA Flow
+
+"auth_req_id", "Auth Result ID" and "Decoupled Auth ID" is version 4 UUID String and it is the key to retrieve the entry from the cache swiftly
+
+These IDs lacks integrity protection and source authentication so that they does not contain any kind of information on CIBA flow.
+
+### Authentication Flow
+
+Like other flows (Browser Flow, Direct Grant Flow, etc.), the authentication flow for CIBA flow is also needed to execute all the processes that need to be executed after the end-user logged-in like creating User Session and Authenticated Client Session that are needed to generate tokens in Token Req/Res part.
+
+However, authentication and authorization is executed by AuthN/AuthZ by AD part so that it is enough for the authentication flow for CIBA flow to contain only the pass-through authenticator. 
+
+### Event
+
+The follwing events are newly added :
+
+* AUTH_REQ_ID_TO_TOKEN : On token endpoint, keycloak successfully generate tokens in return to "auth_req_id".
+* AUTH_REQ_ID_TO_TOKEN_ERROR : On token endpoint, keycloak fails to generate tokens in return to "auth_req_id".
+
+### Server Metadata
+
+The supported server metadata is as follows :
+
+* "backchannel_token_delivery_modes_supported" : ["poll"]
+* "backchannel_authentication_endpoint" : "{scheme}://{host}[:{port}]/auth/realms/{realm}/protocol/openid-connect/backchannelAuthn"
+* "backchannel_user_code_parameter_supported" : false
+
+### Client Metadata
+
+The supported client metadata is as follows :
+
+* "backchannel_token_delivery_mode" : only accepts "poll"
+
+### Token Request Throttling
+
+According CIBA protocol specification, keycloak determines whether it refuses the token request based on its accompanying "auth_req_id".
+
+There are three options considered :
+
+1. Utilize existing `DefaultBruteForceProtector`
+2. Use Inifinispan cache with lifespan of its entry to share among all nodes in the cluster information on which token request with "auth_req_id" is refused
+3. Use some collection with lifespan of its entry on each node, not sharing among all nodes in the cluster information on which token request with "auth_req_id" is refused
+
+1 is too rich for this token request throttling.
+
+2 can satisfies CIBA protocol specification but has performance concern if vast number of token requests happens. When the token request arrives without waiting an interval, keycloak needs to update when this next token request with "auth_req_id" be allowed to be accepted with some additional wait as penalty. Updating the entry of Infinispan cache causes some traffic among nodes in the cluster to sync the status of this entry.
+
+3 can partially satisfies CIBA protocol specification but no traffic is exchanged among nodes in the cluster due to this token request throttling. However, it seems that keycloak does not prepare such collection so that implementing it or introducing some kind of OSS is required.
+
+In this time, 2 is to be realized.
```

Comment:
If the spec could allow issuing the `auth_req_id` every time the client tries to pool tokens from the server we could easily solve this problem. Or even if we could require clients to send an additional parameter for the same purpose so that we know when the last `auth_req_id` was issued and if the time it is being sent is within the expected interval.

I also think #2 is the best way to go (not yet ideal). Although I know how bad the operation can warm a cluster depending on the number of updates.

Reply:
Thank you. At this time, I'll try #2 but considering the performance problem.

Reply:
Yeah, this is a tricky one. I was thinking though ... Why do we need to worry about replicating state across nodes in a cluster but just let each node track `auth_req_id` independently ?

It would be the best effort  and a malicious client wouldn't be able to damage the system so badly.

Reply:
+1 for each node tracking the system independently. Which AFAIK is the solution number 3.

This is good compromise among security and performance IMO. And there is no need to introduce any new infinispan cache into standalone-*.xml file just because that as local cache can be easily created programaticaly. I can imagine something similar to how we create "Revision" caches programatically in the DefaultInfinispanConnectionProviderFactory.initContainerManaged() method.

I've did some small prototype based on the "putIfAbsent" where each item adds the item to the cache with the expiration of interval (5 seconds). The unsuccessful "putIfAbsent" call means that there is already item for this "Auth_req_id" and hence client was too fast with sending request and he should be rejected. The prototype is here: https://github.com/mposolda/keycloak/commit/9918e86e3aedb469405e1e247dee4ba9fc841731

Reply:
@pedroigor @mposolda 
Yes, I agree with you. I'll revise this PR's design document to employ the number 3 method. I'll study how to realize it by studying @mposolda 's prototype.

Reply:
@tnorimat Cool, Thanks!